text/3493-precise-pre-release-cargo-update.md

```diff
@@ -0,0 +1,146 @@
+required by package `tmp-oyyzsf v0.1.0 (/home/ethan/.cache/cargo-temp/tmp-OYyZsF)`
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Consider this table where `a.b.c` is compatible with `x.y.z` and `x.y.z > a.b.c`
+
+| Cargo.toml spec | Cargo.lock version | Target version | Selected by cargo update  | Selected by cargo update --precise  |
+| --------------- | ------------------ | -------------- | ------------------------- | ----------------------------------- |
+| `^a.b.c`         | `a.b.c`            | `x.y.z`        | ✅                        | ✅                                  |
+| `^a.b.c`         | `a.b.c`            | `x.y.z-pre.0`   | ❌                        | ✅                                  |
+| `^a.b.c`         | `x.y.z-pre.0`       | `x.y.z-pre.1`   | ❌                        | ✅                                  |
+| `^a.b.c-pre.0`    | `a.b.c-pre.0`       | `a.b.c-pre.1`   | ✅¹                       | ✅                                  |
+| `^a.b.c-pre.0`    | `a.b.c-pre.0`       | `x.y.z`        | ✅¹                       | ✅                                  |
+| `^a.b.c`         | `a.b.c`            | `a.b.c-pre.0`   | ❌                        | ❌                                  |
+
+✅: Will upgrade
+
+❌: Will not upgrade
+
+¹This behaviour is considered by some to be undesirable and may change as proposed in [RFC: Precise Pre-release Deps](https://github.com/rust-lang/rfcs/pull/3263).
+This RFC preserves this behaviour to remain backwards compatible.
+Since this RFC is concerned with the behaviour of `cargo update --precise` changes to bare `cargo update` made in future RFCs should have no impact on this proposal.
+
+To determine if a version can be selected with `--precise` for a specification that isn't listed above cosider where pre-releases exist within version ranges.
+
+For example consider the version `~1.2.3`.
+The range for `~1.2.3` is [stated in the cargo book](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#tilde-requirements).
+
+```
+~1.2.3  := >=1.2.3, <1.3.0
+```
+
+Intuitively `1.2.4-pre.0` satisfies this inequality, therefore it can be selected with `cargo update --precise`.
+Since it is a pre-release and the specification is not, `1.2.4-pre.0` would not be selected by a bare `cargo update`.
+`1.3.0-pre.0` also satisfies the inequality but `1.2.3-pre.0` and `1.3.1-pre.0` do not.
+
+Put in simple terms the relationship between a pre-release and its stable release is always `a.b.c-pre.0 < a.b.c`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Pre-release versions are not easily auditable when they are only specified in the lock file.
+  A change that makes use of a pre-release version may not be noticed during code review as reviewers don't always check for changes in the lock file.
+- Library crates that require a pre-release version are not well supported since their lock files are ignored by their users (see [future-possibilities])
```

Comment:
From my understand, this has a big drawback that it shifts some "compatible" policy from the central location in the `semver` crate to all callers.

Reply:
I'm not sure what you mean by 'all callers'. Could the semver crate be extended to support the compatible pre-release concept?

Reply:
Maybe @Eh2406 can correct me but I assume the way this would get implement is
- `semver` gets updated to consider pre-releases compatible with a regular version requirement (breaking change)
- All callers of `semver` now need to update to that breaking change and implement on top of `semver` policies for whether pre-releases should be incompatible (e.g. `cargo generate-lockfile`) or not (e.g. `cargo update --precise`).  In these examples, they both rely on the same code paths to make decisions and so we'll have to track the caller of cargo's intentions deep into the call stack to know which policy to select.

Reply:
That largely sounds correct. 

> `semver` gets updated to consider pre-releases compatible with a regular version requirement (breaking change)

We will need to have `requirement.matches(version)` and `requirement.matches_prerelease_flexibility(version)`. I'm not sure where those two functions best live, in the `semver` crate, in cargo, as a conversion on `version`, as a conversion on `requirement`, as a second argument to `matches`, or something else. No matter what there will be two different implementations that need to be used based on context. Whether this is technically a breaking change depends on where it lives and what name we give it.

> All callers of `semver` now need to update to that breaking change and implement on top of semver policies for whether pre-releases should be incompatible (e.g. `cargo generate-lockfile`) or not (e.g. `cargo update --precise`). In these examples, they both rely on the same code paths to make decisions and so we'll have to track the caller of cargo's intentions deep into the call stack to know which policy to select.

I think in cargoes case it will not be too big an imposition. All of these layers of the stack need to keep track of whether a version comes from a lock file for purposes of allowing yanked versions. So the logic goes from `requirement.matches(version)` to `if yank_allow_list.contains(version) {requirement.matches_prerelease_flexibility(version)} else {requirement.matches(version)}`. Gross, but doable.

How this gets adopted in the ecosystem of other tooling around cargo definitely needs to be discussed.